### ISSUE TITLE
A myLeoSettings inconsistency that has cost me too much time

### ISSUE THREAD

**Issue by SegundoBob** [leo-editor/leo-editor#1117]
## The Inconsistency

You can use capital letters in a command name introduced by @command, but you can't use capital letters in a command name invoked by @item.

## The Problem

Suppose you use capital letters in your command introduced by @command. Your command works fine.  When you invoke it from the Minibuffer (Alt-X), you enter the command name in all lower case and the command executes.  show-commands lists the command in all lower case.

Everything is working as intended, so you copy your command name to an @item headline. Now the menu shows your new menu item for your command, but when you pick your new item, nothing happens.  There is no error, but your command is not executed.

## Update: EKR

The underlying problem is that sc.cleanButtonText, in mod_scripting.py, changes the name of commands.  Making a similar change in pbc.doItems appeared to fix the problem, but would be **catastrophically wrong**, because it would clean away Chinese characters in menus!

A proper fix must significantly revise sc.cleanButtonText.  It contains a comment indicating that the cleaning is necessary to make tab completion work.  So fixing this issue will involve the complex code that does tab completion in the minibuffer. 

Furthermore, sc.cleanButtonText disallows upper-case commands defined by `@button` and `@command` nodes.  This surely must be wrong.  Leo already defines vim-related commands that *do* contain upper case letters.

#1121 goes even further and proposes allowing Chinese characters in command names.

The issues are subtle, and span disparate parts of Leo's code base. **Extensive testing will be needed for any proposed fix.** As a result, the only change to the 5.9 code base will be a new warning in k.masterCommand.

## Test System
```
Leo Log Window
Leo 5.9-final-devel, build 20190409061733, Tue Apr  9 06:17:32 UTC 2019
Git repo info: branch = devel, commit = 42b85dd97172
Python 3.6.7, PyQt version 5.11.1
linux
isPython3: True
```

**Comment by maphew** [comment]
Ouch. I probably wouldn't have discovered the reason behind that at all. Good catch, albeit painful one.
Might want to reword subject to match problem more precisely (perhaps "xxxx works in @command but not in @item").

**Comment by edreamleo** [comment]
On Thu, Apr 18, 2019 at 1:04 PM SegundoBob <notifications@github.com> wrote:

You can use capital letters in a command name introduced by @command
> <https://github.com/command>, but you can't use capital letters in a
> command name invoked by @item <https://github.com/item>.


Thanks for this report.  This will be fixed for the b2 release.

Edward

**Comment by edreamleo** [comment]
I have spent some time looking at this. Rev b7ffb57 in devel adds a log message when executing a missing command.  Even this is experimental.

I have not been able to duplicate the problem you report. The following works for me, either in myLeoSettings.leo or in ekr.leo:
```python
@command Test1117
command = 'Test1117'
print(command, command in c.commandsDict, command.lower() in c.commandsDict)
```
Invoking this `@command` node from the minibuffer yields: `Test1117 False True`.

`@item Test1117` also works for me. Executing the command from a menu item gives the same result.

I have looked at the code in detail.  It's not clear why the menu item doesn't work for you.

The check-menu-cmds button in leo-settings.leo will report missing commands.  It used to report that check-leo-file was undefined.  I have disabled the corresponding `@item` node from leoSettings.leo.

I'll leave this item open for now, but I need more information.

**Comment by SegundoBob** [comment]
## What I Did
I updated Leo-editor to ('devel', 'dc846d5556d7').

I put the @command node in myLeoSettings.leo.

I put the @item node in the File menu in leoSettings.leo.

## check-menus-cmds complains
Your last post implies that for you check-menus-cmds does not complain?

When I hit the check-menus-cmds button in leoSettings.leo,  nothing is output to the Log pane.  I consider this a bug, because the following is output to either stdout or stderr and appears on the console:
```
command not found: @menu &File: @item Test1117
done
```
## When the menu item works for me

If the @item node body is empty, then the menu item works and the command is executed---even though check-menus-cmds complains.

## When the menu item does **NOT** work for me
If the @item node body contains "&Test" (i.e., make the menu show "Test" with T as its shortcut), then
hitting the menu item prints this error to the log pane in red:
```
no command for @item 'Test1117'
```

**Comment by edreamleo** [comment]
On Mon, Apr 22, 2019 at 2:48 PM SegundoBob <notifications@github.com> wrote:

> What I Did

I think the problem is the ampersand.  I'll check.

Edward

**Comment by edreamleo** [comment]
Hmm.  I just reread your comment:

QQQ
I put the @command node in myLeoSettings.leo.
I put the @item node in the File menu in leoSettings.leo.
QQQ

It's probably not a great idea to change leoSettings.leo.  I suggest copying the entire `@menu` tree from leoSettings.leo to myLeoSettings.leo.  Please let me know if this solves the problem.

**Comment by SegundoBob** [comment]
I reverted leoSettings.leo to the current released version.
I copied the entire @menu tree from leoSettings.leo to myLeoSettings.leo.
I added the @item node to the File menu in myLeoSettings.leo

The test results are unchanged.

check-menus-cmds complains as long as there is an @item test node.

The test button works when the @item node body is empty and does not work when the @item node body contains "&Test".

**Comment by edreamleo** [comment]
On Mon, Apr 22, 2019 at 6:42 PM SegundoBob <notifications@github.com> wrote:

> I reverted leoSettings.leo to the current released version.
> I copied the entire @menu <https://github.com/menu> tree from
> leoSettings.leo to myLeoSettings.leo.
> I added the @item <https://github.com/item> node to the File menu in
> myLeoSettings.leo
>
> The test results are unchanged.
>
> check-menus-cmds complains as long as there is an @item
> <https://github.com/item> test node.
>
> The test button works when the @item <https://github.com/item> node body
> is empty and does not work when the @item <https://github.com/item> node
> body contains "&Test".
>
Hmm.  Not sure what to do next.  Let me sleep on it.

Edward

**Comment by edreamleo** [comment]
Progress: I can now get a test to fail for me.  The problem likely involves spaces, underscores and '-' symbols in the command name.
```
@command Test1117
@item &test 11-17
```
generates the message `no command for @item '&test 11-17'`

Bob, does this seem relevant to your test case?  I had forgotten to ask the *exact* headlines of your `@command` and `@item` nodes.

**Comment by edreamleo** [comment]
Ok.  I'm sure why the bug happens: c.commandsDict contains "test1117", not "&test 11-17".

This should allow me to fix the bug today.

**Comment by edreamleo** [comment]
Fixed at be21ac in devel.  Please reopen if necessary.

**Comment by edreamleo** [comment]
I am reopening this item, and assigning it for Leo 6.0.  Recent revs, both in "devel" and "py3" back out of the proposed fix.  This faux fix prevents non-unicode characters in `@item` nodes, which would be catastrophic. 

Imo, it is way too late to attempt a fix in the 5.9 code base.  The issues are surprisingly subtle, and mistakes could ruin already existing commands.  The only change that will remain in 5.9 is the new warning in k.masterCommand.

@SegundoBob I see that you are using Python 3, so you should be able to use any fix in the py3 branch.
